coderd/autobuild/lifecycle_executor.go

```diff
@@ -218,53 +200,84 @@ func (e *Executor) runOnce(t time.Time) Stats {
 
-	return false
+	return !currentTick.Before(nextTransition)
 }
 
-func getNextTransition(
-	ws database.Workspace,
-	priorHistory database.WorkspaceBuild,
-	priorJob database.ProvisionerJob,
-) (
-	validTransition database.WorkspaceTransition,
-	nextTransition time.Time,
-	err error,
-) {
-	if !priorJob.CompletedAt.Valid || priorJob.Error.String != "" {
-		return "", time.Time{}, xerrors.Errorf("last workspace build did not complete successfully")
+// isEligibleForAutostart returns true if the workspace should be autostopped.
+func isEligibleForAutostop(build database.WorkspaceBuild, job database.ProvisionerJob, currentTick time.Time) bool {
+	// Don't attempt to autostop failed workspaces.
+	if !job.CompletedAt.Valid || job.Error.String != "" {
+		return false
 	}
 
-	switch priorHistory.Transition {
-	case database.WorkspaceTransitionStart:
-		if priorHistory.Deadline.IsZero() {
-			return "", time.Time{}, xerrors.Errorf("latest workspace build has zero deadline")
-		}
-		// For stopping, do not truncate. This is inconsistent with autostart, but
-		// it ensures we will not stop too early.
-		return database.WorkspaceTransitionStop, priorHistory.Deadline, nil
-	case database.WorkspaceTransitionStop:
-		sched, err := schedule.Weekly(ws.AutostartSchedule.String)
-		if err != nil {
-			return "", time.Time{}, xerrors.Errorf("workspace has invalid autostart schedule: %w", err)
-		}
-		// Round down to the nearest minute, as this is the finest granularity cron supports.
-		// Truncate is probably not necessary here, but doing it anyway to be sure.
-		nextTransition = sched.Next(priorHistory.CreatedAt).Truncate(time.Minute)
-		return database.WorkspaceTransitionStart, nextTransition, nil
-	default:
-		return "", time.Time{}, xerrors.Errorf("last transition not valid for autostart or autostop")
-	}
+	// A workspace must be started in order for it to be auto-stopped.
+	return build.Transition == database.WorkspaceTransitionStart &&
+		!build.Deadline.IsZero() &&
+		// We do not want to stop a workspace prior to it breaching its deadline.
+		!currentTick.Before(build.Deadline)
+}
+
+// isEligibleForFailedStop returns true if the workspace is eligible to be stopped
+// due to a failed build.
+func isEligibleForFailedStop(job database.ProvisionerJob, templateSchedule schedule.TemplateScheduleOptions) bool {
+	// If the template has specified a failure TLL.
+	return templateSchedule.FailureTTL > 0 &&
+		// And the job resulted in failure.
+		db2sdk.ProvisionerJobStatus(job) == codersdk.ProvisionerJobFailed &&
+		// And sufficient time has elapsed since the job has completed.
+		(job.CompletedAt.Valid && database.Now().Sub(job.CompletedAt.Time) > templateSchedule.FailureTTL ||
+			// Or sufficient time has elapsed since the job was canceled.
+			job.CanceledAt.Valid && database.Now().Sub(job.CanceledAt.Time) > templateSchedule.FailureTTL ||
+			// Or the job is stuck/abandoned.
+			database.Now().Sub(job.UpdatedAt) > 30*time.Second)
```

Comment:
This shouldn't be added here and we should rely on my unhanger PR which will be merged soon. Scheduling a new stop job when a job is stuck will result in bigger issues since there will be two active builds for a single workspace.